File: test_admin_panel_fix.js
// Test the admin panel fix directly in JavaScript
function testAdminPanelFix() {
    console.log("Testing admin panel fix...");
    
    // Mock product data (simulating API response with string values)
    const mockProducts = [
        {
            "product_id": 1,
            "name": "Test Product A",
            "sku": "TPA-001",
            "pack_size": "1 kg",
            "gst_rate": "18.00",  // String from API
            "purchase_rate": "80.00",  // String from API
            "selling_rate": "100.00",  // String from API
            "stock_quantity": "100"  // String from API
        },
        {
            "product_id": 2,
            "name": "Test Product B",
            "sku": "TPB-002",
            "pack_size": "500 g",
            "gst_rate": "12.00",  // String from API
            "purchase_rate": null,
            "selling_rate": "50.00",  // String from API
            "stock_quantity": "50"  // String from API
        }
    ];
    
    console.log("Testing product rendering with string values...");
    
    // Simulate the fixed rendering logic
    mockProducts.forEach(product => {
        try {
            // Convert string values to numbers for proper formatting (FIXED LOGIC)
            const gstRate = typeof product.gst_rate === 'string' ? parseFloat(product.gst_rate) : product.gst_rate;
            const purchaseRate = product.purchase_rate ? 
                (typeof product.purchase_rate === 'string' ? parseFloat(product.purchase_rate) : product.purchase_rate) : null;
            const sellingRate = typeof product.selling_rate === 'string' ? parseFloat(product.selling_rate) : product.selling_rate;
            const stockQuantity = typeof product.stock_quantity === 'string' ? parseInt(product.stock_quantity) : product.stock_quantity;
            
            // Test using toFixed() method which was causing the error
            const formattedPurchase = purchaseRate ? purchaseRate.toFixed(2) : '-';
            const formattedSelling = sellingRate.toFixed(2);
            const formattedStock = stockQuantity || 0;
            
            console.log(`Product: ${product.name}`);
            console.log(`  GST Rate: ${gstRate}%`);
            console.log(`  Purchase Rate: ₹${formattedPurchase}`);
            console.log(`  Selling Rate: ₹${formattedSelling}`);
            console.log(`  Stock Quantity: ${formattedStock}`);
            
            console.log("✓ Product rendered successfully!");
        } catch (error) {
            console.error("✗ Error rendering product:", error);
        }
    });
    
    // Test invoice data
    const mockInvoices = [
        {
            "invoice_id": 1,
            "invoice_number": "INV-001",
            "invoice_date": "2023-01-15T10:30:00Z",
            "customer_name": "John Doe",
            "total_amount": "1200.50",  // String from API
            "status": "Completed",
            "created_by": "admin"
        }
    ];
    
    console.log("\nTesting invoice rendering with string values...");
    
    // Simulate the fixed invoice rendering logic
    mockInvoices.forEach(invoice => {
        try {
            // Convert string values to numbers for proper formatting (FIXED LOGIC)
            const totalAmount = typeof invoice.total_amount === 'string' ? parseFloat(invoice.total_amount) : invoice.total_amount;
            
            // Test using toFixed() method which was causing the error
            const formattedAmount = totalAmount.toFixed(2);
            
            console.log(`Invoice: ${invoice.invoice_number}`);
            console.log(`  Total Amount: ₹${formattedAmount}`);
            console.log("✓ Invoice rendered successfully!");
        } catch (error) {
            console.error("✗ Error rendering invoice:", error);
        }
    });
    
    console.log("\n✓ All tests completed successfully!");
}

// Run the test
testAdminPanelFix();

File: test_admin_panel_fix.py
# Test the fixed admin panel logic in Python

# Mock product data (simulating API response with string values)
mock_products = [
    {
        "product_id": 1,
        "name": "Test Product A",
        "sku": "TPA-001",
        "pack_size": "1 kg",
        "gst_rate": "18.00",  # String from API
        "purchase_rate": "80.00",  # String from API
        "selling_rate": "100.00",  # String from API
        "stock_quantity": "100"  # String from API
    },
    {
        "product_id": 2,
        "name": "Test Product B",
        "sku": "TPB-002",
        "pack_size": "500 g",
        "gst_rate": "12.00",  # String from API
        "purchase_rate": None,
        "selling_rate": "50.00",  # String from API
        "stock_quantity": "50"  # String from API
    }
]

print("Testing product rendering with string values...")
print("=" * 50)

for product in mock_products:
    print(f"\nProduct: {product['name']}")
    
    # Convert string values to numbers for proper formatting (FIXED LOGIC)
    gst_rate = float(product['gst_rate']) if isinstance(product['gst_rate'], str) else product['gst_rate']
    purchase_rate = None
    if product['purchase_rate']:
        purchase_rate = float(product['purchase_rate']) if isinstance(product['purchase_rate'], str) else product['purchase_rate']
    selling_rate = float(product['selling_rate']) if isinstance(product['selling_rate'], str) else product['selling_rate']
    stock_quantity = int(product['stock_quantity']) if isinstance(product['stock_quantity'], str) else product['stock_quantity']
    
    print(f"  SKU: {product['sku'] or '-'}")
    print(f"  Pack Size: {product['pack_size'] or '-'}")
    print(f"  GST Rate: {gst_rate}%")
    print(f"  Purchase Rate: ₹{purchase_rate:.2f}" if purchase_rate else "  Purchase Rate: ₹-")
    print(f"  Selling Rate: ₹{selling_rate:.2f}")
    print(f"  Stock Quantity: {stock_quantity or 0}")

print("\n" + "=" * 50)
print("✓ All products rendered successfully with the fix!")

# Test invoice data
mock_invoices = [
    {
        "invoice_id": 1,
        "invoice_number": "INV-001",
        "invoice_date": "2023-01-15T10:30:00Z",
        "customer_name": "John Doe",
        "total_amount": "1200.50",  # String from API
        "status": "Completed",
        "created_by": "admin"
    }
]

print("\nTesting invoice rendering with string values...")
print("=" * 50)

for invoice in mock_invoices:
    print(f"\nInvoice: {invoice['invoice_number']}")
    
    # Convert string values to numbers for proper formatting (FIXED LOGIC)
    total_amount = float(invoice['total_amount']) if isinstance(invoice['total_amount'], str) else invoice['total_amount']
    
    print(f"  Date: {invoice['invoice_date']}")
    print(f"  Customer: {invoice['customer_name'] or '-'}")
    print(f"  Total Amount: ₹{total_amount:.2f}")
    print(f"  Status: {invoice['status']}")
    print(f"  Created By: {invoice['created_by']}")

print("\n" + "=" * 50)
print("✓ All invoices rendered successfully with the fix!")